extensions/amp-story-360/0.1/amp-story-360.js

```diff
@@ -202,15 +272,188 @@ export class AmpStory360 extends AMP.BaseElement {
     container.appendChild(this.canvas_);
     this.applyFillContent(container, /* replacedContent */ true);
 
-    return Services.storyStoreServiceForOrNull(this.win).then(
-      (storeService) => {
+    // Initialize all services before proceeding
+    return Promise.all([
+      Services.storyStoreServiceForOrNull(this.win).then((storeService) => {
+        this.storeService_ = storeService;
+
+        storeService.subscribe(StateProperty.PAGE_SIZE, () =>
+          this.resizeRenderer_()
+        );
+
         storeService.subscribe(
-          StateProperty.PAGE_SIZE,
-          this.resizeRenderer_.bind(this),
-          false /* callToInitialize */
+          StateProperty.GYROSCOPE_PERMISSION_STATE,
+          (permissionState) => this.onPermissionState_(permissionState)
         );
+      }),
+
+      Services.localizationServiceForOrNull(this.element).then(
+        (localizationService) => {
+          this.localizationService_ = localizationService;
+        }
+      ),
+    ]).then(() => {
+      attr('controls') === 'gyroscope' && this.checkGyroscopePermissions_();
+      return Promise.resolve;
+    });
+  }
+
+  /**
+   * @param {string} permissionState
+   * @private
+   */
+  onPermissionState_(permissionState) {
+    if (permissionState === 'granted') {
+      this.enableGyroscope_();
+    } else if (permissionState === 'denied') {
+      this.gyroscopeControls_ = false;
+      this.togglePermissionClass_(true);
+    }
+  }
+
+  /** @private */
+  checkGyroscopePermissions_() {
+    //  Browser doesn't support DeviceOrientationEvent API.
+    if (typeof DeviceOrientationEvent === 'undefined') {
+      return;
+    }
+
+    // If browser doesn't require permission, enable gyroscope.
+    if (typeof DeviceOrientationEvent.requestPermission === 'undefined') {
+      this.enableGyroscope_();
+    }
+
+    // If DeviceOrientationEvent and permissions, build permission button.
+    if (typeof DeviceOrientationEvent.requestPermission === 'function') {
+      this.checkPermissionOnLoad_();
+    }
+  }
+
+  /**
+   * Creates a device orientation listener and sets gyroscopeControls_ state.
+   * If listener is not called in 1000ms, remove listener and resume animation.
+   * This happens on desktop browsers that support deviceorientation but aren't in motion.
+   * @private
+   */
+  enableGyroscope_() {
+    this.gyroscopeControls_ = true;
+    this.togglePermissionClass_(true);
+
+    this.win.addEventListener('deviceorientation', (e) => {
+      if (this.isReady_ && this.isPlaying_) {
+        this.onDeviceOrientation_(e);
       }
+      this.timer_.cancel(checkNoMotion);
+    });
+
+    const checkNoMotion = this.timer_.delay(() => {
+      this.gyroscopeControls_ = false;
+      if (this.isReady_ && this.isPlaying_) {
+        this.animate_();
+      }
+    }, 1000);
+  }
+
+  /**
+   * @param {!Event} e
+   * @private
+   */
+  onDeviceOrientation_(e) {
```

Comment:
Good idea. It's [suggested](https://developers.google.com/web/fundamentals/native-hardware/device-orientation) to sync this to `requestAnimationFrame` so that the frame updates will be smooth.

I created a method called `rafDebounce_` which will only call `onDeviceOrientation_` once per animation frame.

Reply:
This is great!